client/engine/engine.go

```diff
@@ -118,6 +124,11 @@ func (e *Engine) handleMessage(message protocols.Message) ObjectiveChangeEvent {
 
 	e.logger.Printf("Handling inbound message %+v", summarizeMessage(message))
 	objective, err := e.getOrCreateObjective(message)
+	// Acquire a lock for the channels and then refetch the objective with the latest version
+	e.channelLocker.Lock(objective)
+	objective, _ = e.getOrCreateObjective(message)
```

Comment:
The fact that we have to manually refetch the objective after acquiring the lock is not great.